### ISSUE TITLE
degree sequence graph generator (Graph.Degree_Sequence) with “vl” method is super time consuming in python

### ISSUE THREAD

**Issue by yankai-wang** [igraph/python-igraph#413]
**Describe the bug**
There are two equivalent functions in python (Graph.Degree_Sequence) and in R (degree.sequence.game) to generate random graphs. However, when using method = 'vl', the R version has a normal speed while the python version takes a much longer time. For example, the R version takes 27s to generate a graph with 1000 nodes whose degrees are all 750, but the python version takes 839s. I guess there is something wrong since they are supposed to use the same C core?

**To reproduce**
library(igraph)
g=degree.sequence.game(rep(750,1000), method="vl")

import igraph as ig
g = ig.Graph.Degree_Sequence([750]*1000, method='vl')

**Version information**
python-igraph                 0.9.1
R igraph                           1.2.6

**Comment by szhorvat** [comment]
In Mathematica I get 85 seconds on my machine (some of this is the overhead of converting a large graph over to Mathematica format).

**Comment by szhorvat** [comment]
The below C code takes 27 seconds on my machine, while the Python version (with 0.9.4) didn't finish in a few minutes. Something is indeed wrong.

```c
#include <igraph.h>

int main() {
    igraph_t g;
    igraph_vector_t degs;

    igraph_vector_init(&degs, 1000);
    igraph_vector_fill(&degs, 750);

    igraph_degree_sequence_game(&g, &degs, NULL, IGRAPH_DEGSEQ_VL);

    igraph_vector_destroy(&degs);
    igraph_destroy(&g);
    return 0;
}
```

**Comment by ntamas** [comment]
I've done a bit of profiling on macOS on the Python process. It took ~70 seconds for me to generate a graph with 1000 vertices and each vertex having 750 edges from Python and ~33 seconds to do the same thing from C. In the Python profile, it seems like ~15 seconds was spent in `igraph_rng_Python_get_real`, i.e. the part that reaches out to Python's built-in RNG to generate random numbers. That seems to be a low-hanging fruit for optimization so I'll dig a bit deeper into this tomorrow. I can't explain the remaining 22 seconds of difference between Python and C so there's probably also some interesting stuff there.

**Comment by szhorvat** [comment]
> while the Python version (with 0.9.4) didn't finish in a few minutes.

Yesterday I did the measurement incorrectly. I used `%timeit` in IPython, which repeats the commands multiple times, and cancelled the computation before it was done.

Today I get ~65 seconds when measuring correctly, so everything is fine.

----

A more careful measurement in Mathematica gives 69 s with Mathematica's RNG and 39 s with igraph's RNG.  The Python interface can maybe be sped up a bit by providing an option to use igraph's own RNG. However, there are relatively few functions which use the RNG often enough for this to make a difference.

**Comment by ntamas** [comment]
> The Python interface can maybe be sped up a bit by providing an option to use igraph's own RNG.

`set_random_number_generator(None)` is supposed to do this, but it does not work for me; I know where the bug is, but it needs to be fixed first. If I revert to the built-in RNG, the generator completes in 33 seconds, which is comparable to the raw C speed, so it looks like it's definitely the RNG that causes problems here.

**Comment by szhorvat** [comment]
I should also look into making the algorithm interruptible. I'll look into that after the integer transition.

**Comment by ntamas** [comment]
#414 is now fixed, so `set_random_number_generator(None)` will allow one to restore the default igraph RNG, which is faster than Python's RNG due to the fact that it does not have to pass through the Python layer to generate a random number. The fix will be released in 0.9.5.

@yankai-wang If you need this fix before the release, let me know and I'll try to convince the CI system to create a pre-release version with Python wheels that you can use.

**Comment by szhorvat** [comment]
There's still the mystery of why OP sees a > 30x slowdown compared to R while we see a < 3x slowdown, maybe barely 2x.

@yankai-wang Can you let us know:

 - How you timed the computation
 - Where you obtained python-igraph from

**Comment by ntamas** [comment]
I have an unconfirmed theory here. The VL graph generator uses `igraph_rng_get_int31()`, which apparently exists only because the original VL graph generator code used 31-bit random numbers and we did not change that when we vendored the algorithm. `igraph_rng_get_int31()` works by checking whether the `max` parameter of the RNG is 0x7FFFFFFF; if it does, it calls the plain `get()` method of the internal C RNG (which generates a random number between 0 (inclusive) and `max` (exclusive)), otherwise it calls `get_real()` or `get_unif01()` and maps the result into the 0-0x7FFFFFFF range. It is based on the assumption that generating a uniformly distributed random number between 0 and the `max` parameter of the random generator is faster than generating a double in the [0; 1] range and then multiplying it.

Now, the `max` parameter of the RNG of `python-igraph` that maps everything to Python's built-in RNG is set to `LONG_MAX`, which is 0x7FFFFFFF on some platforms and 0x7FFFFFFFFFFFFFFF on others (depending on `sizeof(long int)`). When it happens to be 0x7FFFFFFF, `python-igraph` eventually ends up calling `randint()` from the `random` module, and _not_ `random.random()`, while the latter is faster. (Quite counterintuitive, but the reason is that `random.randint()` jumps through several extra hoops in plain Python).

So, basically, if `sizeof(long int) == 8` on your platform, then you end up on the code branch that calls `random.random()` in Python and you see the ~2x slowdown, while if `sizeof(long int) == 4`, you end up calling `random.randint()` and suffer a performance hit of nearly 30x.

**Comment by ntamas** [comment]
1b2b0a3 now fixes the issue on 32-bit platforms by making sure that the Python RNG always uses `random.random()`. Apparently, `random.randint()` and even `random.getrandbits()` is slower because of the fact that Python does not have "native" integers - every Python integer is a separate `PyLong` object that requires lots of extra bookkeeping.

On the other hand, I am a bit worried that implementing igraph's `get_int31()` on top of `random.random()`, multiplied by 0x7FFFFFFF might not provide a completely uniform distribution. I need to investigate that separately.

**Comment by yankai-wang** [comment]
> There's still the mystery of why OP sees a > 30x slowdown compared to R while we see a < 3x slowdown, maybe barely 2x.
> 
> @yankai-wang Can you let us know:
> 
> * How you timed the computation
> * Where you obtained python-igraph from
> * Can you reproduce it with version 0.9.4

First THANKS a lot! 
I just use time.time() to get the stamps and then compute the difference. I download it with pip and the problem is reproduced with 0.9.4. Below is the result I got.
```
import igraph as ig
import time
s=time.time()
g = ig.Graph.Degree_Sequence([750]*1000, method='vl')
e=time.time()
print(e-s)

1930.3067016601562
```
ps: the difference with my original result may due to different machines rather than different versions. Bust still, this long time is clearly not right.

**Comment by yankai-wang** [comment]
> #414 is now fixed, so `set_random_number_generator(None)` will allow one to restore the default igraph RNG, which is faster than Python's RNG due to the fact that it does not have to pass through the Python layer to generate a random number. The fix will be released in 0.9.5.
> 
> @yankai-wang If you need this fix before the release, let me know and I'll try to convince the CI system to create a pre-release version with Python wheels that you can use.

A pre-release will be perfect! This is the last thing I need to finish my project.

Also, I guess your theory about the random() stuff might be right because I did see something related to it when I forced my machine to stop and in debugging. I forget the details thus cannot reproduce them but I really have the impression.

**Comment by ntamas** [comment]
0.9.5 is now released. This should partially address this issue; the slowdown should now be about 2-3x only, not 30x or more. Furthermore, this release now allows you to call `set_random_number_generator(None)` to make `python-igraph` use igraph's built-in pure C RNG instead of Python's RNG, which should bring the performance to be on par with the C core (but then you can't control the seed of the RNG).
